_posts/2019-03-18-rules-jvm-external-maven.md

```diff
@@ -0,0 +1,142 @@
+rules like `java_import_external` and `aar_import_external`.
+
+## Objectives
+
+We wanted to fulfill two primary objectives with `rules_jvm_external`.
+
+First, users should be able to run `bazel build //...`. This means that the
+calculation of the transitive dependencies and their fetching need to happen
+during repository rule execution time. This makes dependency fetching
+transparent to users and reduces onboarding friction to a Bazel-built project.
+
+Second, we wanted to provide an interface that is familiar to users with
+projects that depend on Maven repositories. The specification of artifacts in
+the repository rule resembles the specification in the `<dependencies>` tag in a
+Maven POM file or the `dependencies` closure in a Gradle build file. In
+particular, this means that users only need to specify their direct dependencies
+and the rule will handle the rest for them.
+
+## Features and use cases
+
+For the initial release, we are targeting support for Java and Android Java
+projects. The following features are supported:
+
+* Supported on Windows, macOS and Linux
+* JAR and AAR packaging types
+* In-band transitive closure resolution
+* Dependency specification directly in the WORKSPACE file
+* Custom Maven repositories
+* Maven repository authentication with HTTP Basic Authentication
+* Artifact version pinning and resolution
+* Multiple independent closures of versioned artifacts
+    * Not forced to only have one version of an artifact per workspace, although
+      different versions are namespaced by the repository name
+* Source JAR support
+* Dependency exclusion
+
+As part of this, we will no longer be maintaining generate_workspace.
+gmaven_rules, which was designed to be an interim solution, will also be
+replaced by rules_jvm_external. We will be updating existing documentation to
+reflect this over the coming weeks.
+
+## How to use
+
+To start using this feature in your Java or Android project, point to the
+repository rule in your WORKSPACE file:
+
+```python
+load("@bazel_tools//tools/build_defs/repo:http.bzl", "http_archive")
+
+http_archive(
+    name = "rules_jvm_external",
+    strip_prefix = "rules_jvm_external-1.1",
+    sha256 = "ade316ec98ba0769bb1189b345d9877de99dd1b1e82f5a649d6ccbcb8da51c1f",
+    url = "https://github.com/bazelbuild/rules_jvm_external/archive/1.1.zip"
+)
+```
+
+Then, specify your direct dependencies using the `artifacts` attribute and
+indicate where those artifacts should come from using the `repositories`
+attribute:
+
+```python
+load("@rules_jvm_external//:defs.bzl", "maven_install")
+
+maven_install(
+    name = "maven",
+    artifacts = [
+        "androidx.test.espresso:espresso-core:3.1.1",
+        "com.google.guava:guava:27.0-android",
+    ],
+    repositories = [
+        "https://maven.google.com",
+        "https://repo1.maven.org/maven2",
+    ],
+    fetch_sources = True,   # Fetch source jars. Defaults to False.
```

Comment:
Why include this in the first example? It feels like the first thing we want to show is the simplest (and perhaps most common) case.

Reply:
I think fetching sources is a very key feature (your IDE won't work if this is set to false). But it makes sense for the default to be false because it is a huge increase in what gets downloaded. So I think that the middle ground of "off by default" but explicitly calling it out in the first example is a good middle ground. WDYT?

Reply:
Fetching source jars alongside your regular jars are a common use case for Maven dependency users.